### ISSUE TITLE
auth=eaponly

### ISSUE THREAD

**Issue by cagney** [libreswan/libreswan#688]
According to the RFC, an IKEv2 EAP exchange looks like:
```
   HDR, SAi1, KEi, Ni  -->
                                <--  HDR, SAr1, KEr, Nr, [CERTREQ]
   HDR, SK {IDi, [CERTREQ,]
       [IDr,] SAi2,
       TSi, TSr}  -->
                                <--  HDR, SK {IDr, [CERT,] AUTH,
                                         EAP}
   HDR, SK {EAP}  -->
                                <--  HDR, SK {EAP (success)}
   HDR, SK {AUTH}  -->
                                <--  HDR, SK {AUTH, SAr2, TSi, TSr}
```
does eaponly mean that the last exchange is omitted?

**Comment by cagney** [comment]
See https://datatracker.ietf.org/doc/html/rfc5998 I guess

**Comment by cagney** [comment]
After ca9a9e8a7420b13bfc16656415353c109ffe636f and e828c566de991bced38cac49f7db44f0685346c6
```
"rw-eap"[2] 192.1.3.209 #2: sent IKE_SA_INIT reply {cipher=AES_GCM_16_256 integ=n/a prf=HMAC_SHA2_512 group=MODP2048}
"rw-eap"[2] 192.1.3.209 #2: processing decrypted IKE_AUTH request: SK{IDi,N(INITIAL_CONTACT),CERTREQ,IDr,CP,SA,TSi,TSr,N(MOBIKE_SUPPORTED),N(NO_ADDITIONAL_ADDRESSES),N(EAP_ONLY_AUTHENTICATION),N(IKEV2_MESSAGE_ID_SYNC_SUPPORTED)}
"rw-eap"[2] 192.1.3.209 #2: ignoring CERTREQ payload that is not ASN1: content is not binary ASN.1
"rw-eap"[2] 192.1.3.209 #2: added EAP payload to packet
"rw-eap"[2] 192.1.3.209 #2: NSS: I/O getpeername
"rw-eap"[2] 192.1.3.209 #2: sent EAP request
"rw-eap"[2] 192.1.3.209 #2: EAP TLS Msglen 181
"rw-eap"[2] 192.1.3.209 #2: EAP with 181 bytes, flags 80
"rw-eap"[2] 192.1.3.209 #2: NSS: I/O getsocketoption(Nonblocking)
"rw-eap"[2] 192.1.3.209 #2: NSS: I/O recv(5): return 5
"rw-eap"[2] 192.1.3.209 #2: NSS: I/O recv(176): return 176
"rw-eap"[2] 192.1.3.209 #2: NSS: I/O getpeername
"rw-eap"[2] 192.1.3.209 #2: NSS: I/O send(3824): ok
"rw-eap"[2] 192.1.3.209 #2: NSS: I/O recv(5): would block
"rw-eap"[2] 192.1.3.209 #2: Responding with 1014 bytes of 3824 EAP data
"rw-eap"[2] 192.1.3.209 #2: error initializing padding for encrypted ISAKMP Message payload: IKEv2 Encryption Payload is already full; discarding 1 0x00 repeated bytes of padding and length
"rw-eap"[2] 192.1.3.209 #2: state transition function for STATE_V2_PARENT_R_EAP had internal error
```

**Comment by cagney** [comment]
After fixing 31fb7d6820d
```
"rw-eap"[2] 192.1.3.209 #2: NSS: I/O send(31): ok
"rw-eap"[2] 192.1.3.209 #2: NSS: I/O close
"rw-eap"[2] 192.1.3.209 #2: authentication failed: computed hash does not match hash received from peer ID_IPV4_ADDR '192.1.3.209'
```

**Comment by cagney** [comment]
This:
```
		DBG_dump_hunk("PSK", pss);
		DBG_dump_hunk("idhash", *idhash);
		DBG_dump_hunk("PSK", pss);
```
where DBG_dump_hunk() is defined as:
```
#define DBG_dump_hunk(LABEL, HUNK)				\
	{							\
		typeof(HUNK) hunk_ = HUNK; /* evaluate once */	\
		DBG_dump(LABEL, hunk_.ptr, hunk_.len);		\
	}
```
(notice how it is copying `*idhash` onto the stack) prints:
```
| PSK
|   89 89 71 64  49 1f 94 85  b7 ea 14 6a  b4 b8 b3 ac   ..qdI......j....
|   7f 82 7c c0  0b e9 eb cd  b2 c1 3e 57  f0 63 17 5f   ..|.......>W.c._
|   4d 06 49 27  0a 2a 57 c7  0e c9 ef 65  82 d9 e8 fc   M.I'.*W....e....
|   f8 ad 25 52  13 90 fd f2  67 6a e6 85  37 3f 63 e7   ..%R....gj..7?c.
| idhash
|   71 b0 6d 55  49 8c 8c f6  8f 1e e9 2a  a1 9b 3a 8b   q.mUI......*..:.
|   4c 2a 9d 93  95 8a 51 a1  a8 1e ae 20  f3 84 0d 25   L*....Q.... ...%
|   fe 3e 67 55  ce 67 23 0a  91 f1 a3 ea  92 cd 27 20   .>gU.g#.......' 
|   4a 2e ef cc  c0 6c 3f 16  9a 33 25 a7  1c f1 14 5d   J....l?..3%....]
| PSK
|   71 b0 6d 55  49 8c 8c f6  8f 1e e9 2a  a1 9b 3a 8b   q.mUI......*..:.
|   4c 2a 9d 93  95 8a 51 a1  a8 1e ae 20  f3 84 0d 25   L*....Q.... ...%
|   fe 3e 67 55  ce 67 23 0a  91 f1 a3 ea  92 cd 27 20   .>gU.g#.......' 
|   4a 2e ef cc  c0 6c 3f 16  9a 33 25 a7  1c f1 14 5d   J....l?..3%....]
```
this is not good.

(if the macro is changed to use pointers and some other code is tweaked then the responder authenticates)

**Comment by cagney** [comment]
This seems to better match strongswan which only uses 64-bits returned by SSL_ExportKeyingMaterial()
```
@@ -631,7 +631,8 @@ stf_status process_v2_IKE_AUTH_request_EAP_final(struct ike_sa *ike,
 	struct eap_state *eap = ike->sa.st_eap;
 	struct msg_digest *sa_md = ike->sa.st_eap_sa_md;
 	struct logger *logger = ike->sa.st_logger;
-	struct hash_signature msk = { .len = sizeof(msk.ptr/*array*/), };
+	struct hash_signature msk = { .len = 64/* from RFC? */, };
+	passert(msk.len <= sizeof(msk.ptr/*array*/));
 
 	pexpect(eap != NULL);
 	pexpect(sa_md != NULL);
```
the result (ignoring the potential compiler issue) is that the responder authenticates; but the initiator then fails

**Comment by cagney** [comment]
This fixes the MSK size problem 0a2dbc9cd7750c68c1422b570712dc976e6244b1

**Comment by letoams** [comment]
No, but the AUTH is changed, https://datatracker.ietf.org/doc/rfc5998/

Sent using a virtual keyboard on a phone
